rfcs/20200612-stream-executor-c-api.md

```diff
@@ -0,0 +1,557 @@
+  // here.
+  void* device_handle;
+
+  // Any kind of data that plugin device might want to store.
+  void* data;
+} SE_Device;
+
+#define SE_DEVICE_STRUCT_SIZE TF_OFFSET_OF_END(SE_Device, data)
+
+typedef struct SE_StreamExecutor {
+  size_t struct_size;
+  void* ext;
+
+  /*** ALLOCATION CALLBACKS ***/
+  // Synchronously allocates size bytes on the underlying platform and returns
+  // a DeviceMemoryBase representing that allocation. In the case of failure,
+  // nullptr is returned.
+  TF_DeviceMemoryBase* (*allocate)(
+      SE_Device* se, uint64_t size, int64_t memory_space);
+
+
+  // Deallocate the DeviceMemory previously allocated via this interface.
+  // Deallocation of a nullptr-representative value is permitted.
+  void (*deallocate)(
+      SE_Device* se, TF_DeviceMemoryBase* memory);
+
+
+  // Return allocator statistics.
+  TF_BOOL (*get_allocator_stats)(SE_Device* executor,
+                              SE_AllocatorStats* stats);
+  // Returns the underlying device memory usage information, if it is available.
+  // If it is not available (false is returned), free/total may not be
+  // initialized.
+  TF_BOOL (*device_memory_usage)(
+      SE_Device* executor, int64_t* free, int64_t* total);
+
+
+  /*** STREAM CALLBACKS ***/
+  // Creates SE_Stream. This call should also Allocate stream
+  // resources on the underlying platform and initializes its
+  // internals.
+  void (*create_stream)(SE_Device* executor, SE_Stream*, TF_Status*);
+
+  // Destroys SE_Stream and deallocates any underlying resources.
+  void (*destroy_stream)(SE_Device* executor, SE_Stream stream);
+
+  // Causes dependent to not begin execution until other has finished its
+  // last-enqueued work.
+  TF_BOOL (*create_stream_dependency)(
+      SE_Device* executor, SE_Stream dependent,
+      SE_Stream other);
+
+  // Without blocking the device, retrieve the current stream status.
+  void (*get_status)(SE_Device* executor, SE_Stream stream,
+                     TF_Status* status);
+
+  /*** EVENT CALLBACKS ***/
+  // Create SE_Event. Performs platform-specific allocation and initialization of an event.
+  void (*create_event)(
+      SE_Device* executor, SE_Event* event, TF_Status* status);
+
+  // Destroy SE_Event and perform any platform-specific deallocation and cleanup of an event.
+  void (*destroy_event)(
+      SE_Device* executor, SE_Event event, TF_Status* status);
+
+  // Requests the current status of the event from the underlying platform.
+  SE_EventStatus (*poll_for_event_status)(
+      SE_Device* executor, SE_Event event);
+  // Inserts the specified event at the end of the specified stream.
+  void (*record_event)(
+      SE_Device* executor, SE_Stream stream,
+      SE_Event event, TF_Status* status);
+
+  // Wait for the specified event at the end of the specified stream.
+  void (*wait_for_event)(
+      SE_Device* executor, SE_Stream stream,
+      SE_Event event, TF_Status* status);
```

Comment:
Do we need an API to block the host until an event has completed?  If we have it then we won't technically need `block_host_until_done` since that can be implemented by enqueueing an event on a stream and blocking on that event.

Reply:
Sounds reasonable to me (as I understand `wait_for_event` is such an API). May be we can remove `block_host_until_done`.
I suspect the only issue might arise if the plugin does not support events.

@jzhoulon do you think just having `wait_for_event` and no `block_host_until_done` would be enough for PluggableDevice?

Reply:
No, I think these two api has different semantics, wait_for_event is an async behavior, that is stream wait for an event from another stream(). that is this API deals with synchronization between multiple streams:
```
Stream &Stream::ThenWaitFor(Event *event) {
    port::Status status = parent_->WaitForEvent(this, event);
```
 blocking_host_until_done is an sync behavior, that is cpu wait for the GPU result, a blocking behavior

Reply:
I agree that just `wait_for_event` is not sufficient.  It only inserts a "wait for event" operation in the relevant stream.  I'm suggesting the equivalent of `cudaEventSynchronize` which can be used to implement `cudaStreamSynchronize` (with a minor performance penalty).

Reply:
We could consider it, but it hasn't been on the table. Right now we were keeping StreamExecutor API the same and just wrapping it with C API. This sounds like a StreamExecutor API change.

Reply:
Based on design review, seems like we do want a way for a host to wait for event.

Related question: how useful is current `wait_for_event`? Should we keep it as well?
This way we will have two functions `wait_for_event` and `host_wait_for_event` (may be there is a better name).

Reply:
> Related question: how useful is current wait_for_event? Should we keep it as well?

Very useful. :). That is how we create dependencies between streams.

Probably we should s/`host_wait_for_event`/`block_host_for_event`/?

Reply:
may be host_block_for_event?

Reply:
Thank you for all the info!

Added:
```cpp
  // Causes the host code to synchronously wait for the event to complete.
  void (*block_host_for_event)(
      SP_Device* executor, SP_Event event, TF_Status* status);
```

I also removed `block_host_until_done` per Sanjoy's original suggestion. Let me know if we need both.